File: Serialize and Deserialize a Binary Tree.py
# Time:- O(n)
# Space:- O(n)

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.
        
        :type root: TreeNode
        :rtype: str
        """
        # BFS to traverse and serialize the tree
        if not root:
            return ""
        q=collections.deque([])
        q.append(root)
        string=""
        while(q):
            ele=q.popleft()
            if ele=="null":
                string+=ele+" "
                continue
            else:
                string+=str(ele.val)+" "
            if ele.left:
                q.append(ele.left)
            else:
                q.append("null")
            if ele.right:
                q.append(ele.right)
            else:
                q.append("null")
        return string
                
            
    def deserialize(self, data):
        """Decodes your encoded data to tree.
        
        :type data: str
        :rtype: TreeNode
        """
        # put it back by iterating on the nodes
        nodes=data.strip().split(" ")
        if nodes==[""]:
            return None
        nodes[0]=TreeNode(int(nodes[0]))
        i,j=0,1
        while(j<len(nodes)):
            if nodes[i]=="null":
                i+=1
                continue
            if j<len(nodes) and nodes[j]!="null":
                nodes[j]=TreeNode(int(nodes[j]))
                nodes[i].left=nodes[j]
            if j+1<len(nodes) and nodes[j+1]!="null":
                nodes[j+1]=TreeNode(int(nodes[j+1]))
                nodes[i].right=nodes[j+1]
            j+=2
            i+=1
        return nodes[0]
            
        
        

# Your Codec object will be instantiated and called as such:
# ser = Codec()
# deser = Codec()
# ans = deser.deserialize(ser.serialize(root))

File: SerializeAndDeserializeABinaryTree.java


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        String retStr = "";
        if(root == null) return retStr;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while(!queue.isEmpty()){
            TreeNode front = queue.poll();
            
            if(front == null){
                retStr += "NULL,";
            } else{
                retStr += Integer.toString(front.val)+ ",";
                queue.add(front.left);
                queue.add(front.right);
            }
            // we need NUll as well so no need of for loop
          
        }
        while(!Character.isDigit(retStr.charAt(retStr.length()-1))){
            retStr = retStr.substring(0,retStr.length()-1);
        }
        
        return retStr;
        
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
     if(data == "") return null;
       
    String[] processed = data.split(",");
    int index = 0;
        TreeNode root = new TreeNode(Integer.parseInt(processed[index]));
        index+= 1;
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        while(!queue.isEmpty() && index < processed.length){
            TreeNode front = queue.poll();
            
            if(index < processed.length && processed[index].compareTo("NULL") !=0){
                front.left = new TreeNode(Integer.parseInt(processed[index]));
                queue.add(front.left);
            }
            index+=1;
            
            if(index < processed.length && processed[index].compareTo("NULL") !=0){
                front.right = new TreeNode(Integer.parseInt(processed[index]));
                queue.add(front.right);
            }
            index+=1;
        }
        return root;
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));
